### DISCUSSION TITLE
Chakra Styles not being applied in Storybook

### DISCUSSION THREAD

**Discussion by ntwrkjasen** [chakra-ui/chakra-ui#5987] [Q&A]
Hello, I've been banging my head against the wall for a few days. I'm using the official chakra ui storybook add on and can't get the chakra styles to be applied to my storybook components. Any help would be MUCH appreciated.

main.js

```
module.exports = {
  stories: ['../stories/**/*.stories.mdx', '../stories/**/*.stories.@(js|jsx|ts|tsx)'],
  addons: [
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    '@storybook/addon-interactions',
    '@chakra-ui/storybook-addon',
  ],
  framework: '@storybook/react',
  core: {
    builder: 'webpack5',
  },
  features: {
    emotionAlias: false,
  },
};
```

previews.js

```
import { extendTheme } from '@chakra-ui/react';
const theme = require('../src/lib/styles/theme');

const config = {
  useSystemColorMode: false,
};

const customTheme = extendTheme({ config, ...theme });

export const parameters = {
  chakra: { customTheme },
  actions: { argTypesRegex: '^on[A-Z].*' },
  controls: {
    matchers: {
      color: /(background|color)$/i,
      date: /Date$/,
    },
  },
};
```

theme.tsx

```
const breakpoints = createBreakpoints({
  lg: '1024px',
  md: '768px',
  sm: '425px',
  xl: '1440px',
});

export default extendTheme({
  borders: {
    brand: {
      1: '1px solid #000000',
      2: '1px solid #333333',
      3: '1px solid #666666',
      4: '1px solid #999999',
      5: '1px solid #CCCCCC',
      6: '1px solid #E6E6E6',
      7: '1px solid #F9F9F9',
      8: '1px solid #FFFFFF',
    },
    error: {
      1: '1px solid #e53e3e',
      2: '2px solid #e53e3e',
      3: '3px solid #e53e3e',
    },
  },
  breakpoints,
  colors: {
    admin: '#41E051',
    brand1: '#000000',
    brand2: '#333333',
    brand3: '#666666',
    brand4: '#999999',
    brand5: '#CCCCCC',
    brand6: '#E6E6E6',
    brand7: '#F9F9F9',
    brand8: '#FFFFFF',
    brand9: '#2136F6',
    error: '#e53e3e',
    ntwrkBlack: {
      // custom black that can be used in color scheme props
      50: '#666666',
      100: '#666666',
      200: '#666666',
      300: '#666666',
      400: '#666666',
      500: '#000000',
      600: '#000000',
      700: '#000000',
      800: '#000000',
      900: '#000000',
    },
    success: '#76D22E',
  },
  components: {
    CircularImage,
    Popover: {
      variants: {
        gridUserNames: {
          popper: {
            marginTop: '0 !important',
            maxWidth: 'none',
            width: 'auto',
          },
        },
      },
    },
    Table: {
      baseStyle: {
        th: {
          textTransform: 'none',
        },
      },
      parts: ['th'],
      variants: {
        simple: {
          td: {
            fontSize: '.75rem',
          },
          th: {
            borderBottom: 'none',
            color: 'brand2',
            fontSize: '0.813rem',
            fontWeight: '600',
          },
        },
      },
    },
  },
  fontSizes: {
    '2xl': 18 / 16 + 'rem',
    '3xl': 20 / 16 + 'rem',
    '4xl': 24 / 16 + 'rem',
    '5xl': 28 / 16 + 'rem',
    '6xl': 40 / 16 + 'rem',
    lg: 13 / 16 + 'rem',
    md: 12 / 16 + 'rem',
    sm: 11 / 16 + 'rem',
    xl: 16 / 16 + 'rem',
    xs: 10 / 16 + 'rem',
  },
  fontWeights: {
    medium: 500,
    normal: 400,
  },
  fonts: {
    body: 'Graphik',
    heading: 'Graphik',
  },
  letterSpacings: {
    wide: '0.2px',
    wider: '0.3px',
  },
  lineHeights: {
    base: '20px',
    normal: '22px',
    short: '17px',
    shorter: '16px',
    shortest: '15px',
    tall: '24px',
    taller: '28px',
    tallest: '32px',
  },
  shadows: {
    box: '0 0 10px rgba(0,0,0,0.25)',
  },
  sizes: {
    '13': '3.25rem',
    '15': '3.75rem',
    '50': '12.5rem',
    '65': '16.25rem',
    '75': '18.75rem',
  },
});
```

**Comment by ntwrkjasen** [comment]
This can be closed. I'm dumb 🤦🏾‍♂️

**Reply by oscarsala** [reply]
what was your solution?

**Reply by fredrik-sogaard** [reply]
Yes, curious as well! Could you post your solution?

**Reply by jovangacic** [reply]
maybe it was this the fact that he was passing `customTheme` and needed to pass `theme`

**Comment by gabrielfgularte** [comment]
I'm facing the same problem, but in my case, I think that I'm placing the custom theme in a right way.

main.js
```
const path = require('path')

module.exports = {
  stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
  addons: [
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    '@storybook/addon-interactions',
    '@storybook/preset-create-react-app',
    '@chakra-ui/storybook-addon',
  ],
  framework: '@storybook/react',
  core: {
    builder: '@storybook/builder-webpack5',
  },
  features: {
    emotionAlias: false,
  },
    return config
  },
}
```

preview.js
```
const theme = require('../src/theme')

export const parameters = {
  chakra: { theme },
  actions: { argTypesRegex: '^on[A-Z].*' },
  controls: {
    matchers: {
      color: /(background|color)$/i,
      date: /Date$/,
    },
  },
}
```

All the components appear, but they don't get my custom theme. =/ @ntwrkjasen can you explain how have you managed to solve that? It can help me out on this one. Thanks.

**Reply by gabrielfgularte** [reply]
@JovanGacic I see that you have commented before with a possible solution. Do you think that I'm doing something wrong? Thanks.

**Reply by jovangacic** [reply]
@gabrielfgularte looks ok from here, do you have a branch/example to share to check it out?

**Reply by gabrielfgularte** [reply]
Hi there, sorry I was into other many things in my project. But here it is: https://github.com/gabrielfgularte/chakra-storybook

I created a react app with CRA, then added chakra and then storybook. Ran into these steps: https://chakra-ui.com/getting-started/with-storybook, created my own theme (`src/theme.js`) but when I run `npm run storybook`, the chakra button doesn't get my custom theme applied, although you can see it working as expected at `localhost:3000` below the "Learn React" link.

**Reply by 60pfennig** [reply]
hi @gabrielfgularte I guess your problem is that you are using the examples of storybook or do I miss something? So in your stories not the Chakra Button is used, but the Storybook example ones.
If you mean the addon generated ones, I'm not sure if they are meant to be styled by your theme. It would be convinient tho. For me I just created a story for the Chakra Button ad it works:
```
//styledButton.stories.tsx
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Button } from '@chakra-ui/react';

export default {
  title: 'Button/StyledButton',
  component: Button,

} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args}>Anmelden</Button>;

export const Primary = Template.bind({});
Primary.args = {
  size: 'm',
  variant: "solid"
};

```
also make sure you are applying @dyelax suggestion and changeing require(...) to `import theme from THEMEPATH`

**Reply by gabrielfgularte** [reply]
No, actually the storybook examples aren't changed by my custom theme at all. What I expect is to see these components (a list with Chakra components) with my custom theme applied to each one:

![image](https://user-images.githubusercontent.com/1307137/217607150-cc5b955b-13a3-41d5-b5f0-c670d7575dbe.png)

**Reply by gabrielfgularte** [reply]
And I've tried both with require and import. It doesn't work either way.

**Reply by 60pfennig** [reply]
yes I think this is not intendet to work or its a bug in the ChakraUI storybook addon. If you create the an own story for each component it will work.

**Reply by pawangujral** [reply]
Chakra Storybook won't change as it is a vendor storybook, wrapped for you just as an example. You have to create your custom component if you want to see your changes.

**Reply by pawangujral** [reply]
Not recommend it, But if you really want to do that, you need to import your theme in Chakra package within `node_module`. and keep patch sync so you don't lose your custom changes. bit hard but doable, if you dig deep :) .. I never tried it but I'm hopeful it will work.

**Reply by ivorynoise** [reply]
@gabrielfgularte did you manage to override the default theme for components under Chakra UI section in storybook ? I have exact requirements and cannot figure out a way except patching node_modules.

**Comment by pawangujral** [comment]
for rest of fellow searcher who are still banging their head for solution, this is how I solved it understanding from this thread.
 
Add this in preview.js
```
const theme = extendTheme({
  colors: {
    brand: 'green', // Whatever your custom changes are
  },
});

export const parameters = {
  chakra: { theme }, // THIS <===
  .......
}
```
In Component

` <Button bg="brand">
      Click me
    </Button>
`

Hope it helps :)

**Reply by 60pfennig** [reply]
Hi all, 
I don't get this to work either. The global styles are applied in storybook, but custom components are not styled.
I exactly using the documentation from the docs.
```
module.exports = {
  stories: [
    "../stories/**/*.stories.mdx",
    "../stories/**/*.stories.@(js|jsx|ts|tsx)",
  ],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
    "@chakra-ui/storybook-addon",
  ],
  framework: "@storybook/react",
  core: {
    builder: "@storybook/builder-webpack5",
  },
  features: { emotionAlias: false },
  webpackFinal: async (config) => {
    config.module.rules.push({
      test: /\.mjs$/,
      include: /node_modules/,
      type: "javascript/auto",
    });
    return config;
  },
};
```
and preview:
```
const theme = require('./../constants/theme')

export const parameters = {
  actions: { argTypesRegex: "^on[A-Z].*" },
  chakra: {
    theme: {...theme},
    
  },
  controls: {
    matchers: {
      color: /(background|color)$/i,
      date: /Date$/,
    },
  },
}`
My theme:
`
export const theme = extendTheme({
  useSystemColorMode: false,
  fonts: {
    body: "Inter open-sans arial",
    heading: "Inter open-sans arial",
  },
  colors: {
    primary: {
      500: "#0D9869",
      600: "#057A55",
    },
  },
  components: {
    Button: { ...Button },
  },
});

export default theme
```
The font is changing so I guess the theme is somhow applied.
But the button style is not changed (changing the background). It is fully working in my Next.js application so the theme should be valid. But in Storybook the buttons background stays gray.
I'm getting several warnings/errors in dev tools tho:

```
Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot 

Quellübergreifende (Cross-Origin) Anfrage blockiert: Die Gleiche-Quelle-Regel verbietet das Lesen der externen Ressource auf https://storybook.chakra-ui.com/metadata.json?version=2.4.9. (Grund: CORS-Kopfzeile 'Access-Control-Allow-Origin' fehlt). Statuscode: 404.

You are loading @emotion/react when it is already loaded. Running multiple instances may cause problems. This can happen if multiple versions are used, or if multiple builds of the same version are used.

Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot

Source-Map-Fehler: Error: request failed with status 404
Ressourcen-Adresse: http://localhost:6006/vendors-node_modules_chakra-ui_storybook-addon_dist_preset_register_js-node_modules_storybook-46c2aa.manager.bundle.js
Source-Map-Adresse: index.js.map
```
thank you for your help <3

**Reply by pawangujral** [reply]
@60pfennig

 If this is only for button. `colorScheme` prop only take color object. example below. you need to generate entire object to make it work not partial. e.g Use this https://smart-swatch.netlify.app/ (type your color and copy js object) put it in your theme object. It will work just fine. :)

```
   colors: {
    brand: { 
        primary: {
          50: '#ffe2e2',
          100: '#ffb1b2',
          ........
          800: '#4f0000',
          900: '#200000',
        }
}
}
```

**Reply by 60pfennig** [reply]
Hi @pawangujral thanks for your reply :)
this sadly does't resolve the issue. I added the missing values but this doesn't change anything in my storybook. I forgot to mention that the button is styled correctly in my Next.js app. In Storybook it just takes the right font (the global style) for the button but it don't get styled according to my custom Button styles. If I'm using my own variants (e.g. solid-primary) which aren't  used by Chakra it will displayed just unstyled.

**Reply by gabrielfgularte** [reply]
Hey @pawangujral it doesn't work for me either. I was already with the `chakra: {theme}` stuff in y `preview.js` and it didn't work. Please, refere to[ this comment](https://github.com/chakra-ui/chakra-ui/discussions/5987#discussioncomment-4895184) to check the repo I've created to expose the issue. Thanks.

**Reply by pawangujral** [reply]
@gabrielfgularte cloned your repo and fixed it (In whatever way I understood your problem). If you can allow my account to push code to your repos, I can push new branch for your to check.

**Reply by pawangujral** [reply]
@60pfennig Share your repo, will check.

**Reply by gabrielfgularte** [reply]
> @gabrielfgularte cloned your repo and fixed it (In whatever way I understood your problem). If you can allow my account to push code to your repos, I can push new branch for your to check.

@pawangujral  Sure. Done! Thank you!

**Reply by pawangujral** [reply]
Created draft MR  by that example, you can customise each one. Let me know if I misunderstood your requirement and doesn't work.

**Reply by 60pfennig** [reply]
Hi @pawangujral, for me the solution below from @dyelax worked.
I changed `const theme = require("/styles/theme");` to `import theme from "../styles/theme";`
Note: don't use curly braces: ~~`import {theme} from "../styles/theme";`~~

**Reply by pawangujral** [reply]
Awesome. 👍🏼

**Comment by dyelax** [comment]
I ran into a similar problem and it was because my theme wasn't importing correctly into `.storybook/preview.js`. Following the setup instructions, I had `const theme = require("/styles/theme");`, but I had to change it to `import theme from "../styles/theme";`

Note the 2 changes:
* Change from `require` => `import`
* Change from absolute path to relative path

**Reply by gabrielfgularte** [reply]
It doesn't work for me. Even if I put the theme on `storybook/preview.js` without import it, like this:
```
# storybook/preview.js

import { extendTheme } from '@chakra-ui/react'
const theme = extendTheme({
  components: {
    Button: {
      baseStyle: {
        bg: 'blue',
        backgroundColor: 'blue',
        color: 'white',
      }
    }
  }
})

export const parameters = {
  chakra: { theme },
  actions: { argTypesRegex: '^on[A-Z].*' },
  controls: {
    matchers: {
      color: /(background|color)$/i,
      date: /Date$/,
    },
  },
}
```

Please refer to [this comment](https://github.com/chakra-ui/chakra-ui/discussions/5987#discussioncomment-4895224) - I've created a repo to expose the issue.

**Comment by gabrielfgularte** [comment]
None of the suggestions above have worked, so I created a repo to debug the problem: https://github.com/gabrielfgularte/chakra-storybook. Please, refer to [this comment](https://github.com/chakra-ui/chakra-ui/discussions/5987#discussioncomment-4895184) to see what I did.

Although you can see the custom themed button working on `localhost:3000`, the storybook panel doesn't render the themed button. It keeps the basic chakra style from default theme.

**Reply by gabrielfgularte** [reply]
@ntwrkjasen if you have managed to solve this issue, please, can you illustrate what you did? Thank you!

**Comment by wazcov** [comment]
For me the problem was to change this:

`const preview: Preview = {
  parameters: {
    chakra: {
      CustomTheme,
    },
    actions: { argTypesRegex: "^on[A-Z].*" },
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/i,
      },
    },
  },
};`

to this

`const preview: Preview = {
  parameters: {
    chakra: {
      theme: CustomTheme,
    },
    actions: { argTypesRegex: "^on[A-Z].*" },
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/i,
      },
    },
  },
};`

Note how under chakra, you have to specify a `theme: <importOfYourTheme>`

**Comment by ilhamkukuh** [comment]
am I also experiencing the same issue? All the answers above did not work. Any updates for this?

I use storybook ^7.5.3

**Comment by Rawallon** [comment]
Hey, I found the cause and a workaround:

The problem is that configs on Preview.ts are not getting loaded, I don't know why, I'll leave that for someone else to figure it out, once I noticed that figuring out a workaround was pretty quick.

I'm using [Decorators](https://storybook.js.org/docs/writing-stories/decorators) to wrap the component on ChakraProvider, here's an example story (also, don't forget to rename the extension from .ts to .tsx)

```js
import type { StoryObj } from '@storybook/react';
import { Button, ChakraProvider } from '@chakra-ui/react';
import { extendedTheme } from 'ui';

const meta = {
  title: 'Example/Button',
  component: Button,
  parameters: {
    layout: 'centered',
  },
  decorators: [
    (Story: any) => (
      <ChakraProvider theme={extendedTheme}>
        <Story />
      </ChakraProvider>
    ),
  ],
};


export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    bg: 'blue',
    children: 'Keep on keeping on',
  },
};
```

**Reply by jude98** [reply]
Finally this works!!
But I think instead of wrapping each stories, we can add decorators in preview (global decorators):

```typescript

import React from 'react';
import { ChakraProvider } from '@chakra-ui/react';
import type { Preview } from '@storybook/react';

import theme from 'path_to_theme';

const preview: Preview = {
  decorators: [
    (Story) => (
      <ChakraProvider theme={theme}>
        <Story />
      </ChakraProvider>
    ),
  ],
  parameters: {
    chakra: {
      theme,
    },
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/i,
      },
    },
  },
};

export default preview;

```

P.S:  make sure to name the file as tsx or jsx

**Reply by Rawallon** [reply]
Awesome to hear, I honestly just dropped Storybook after that feature was completed
